test/OpenTelemetry.Extensions.Hosting.Tests/OpenTelemetryMetricsBuilderExtensionsTests.cs

```diff
@@ -0,0 +1,167 @@
+using Microsoft.Extensions.Configuration;
+using Microsoft.Extensions.Configuration.Memory;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Diagnostics.Metrics;
+using Microsoft.Extensions.Options;
+using OpenTelemetry.Internal;
+using OpenTelemetry.Logs;
+using OpenTelemetry.Metrics;
+using OpenTelemetry.Metrics.Tests;
+using OpenTelemetry.Tests;
+using OpenTelemetry.Trace;
+using Xunit;
+
+namespace OpenTelemetry.Extensions.Hosting.Tests;
+
+public class OpenTelemetryMetricsBuilderExtensionsTests
+{
+    [Theory]
+    [InlineData(false)]
+    [InlineData(true)]
+    public void EnableMetricsTest(bool useWithMetricsStyle)
+    {
+        using var meter = new Meter(Utils.GetCurrentMethodName());
+        List<Metric> exportedItems = new();
+
+        using (var host = MetricTestsBase.BuildHost(
+            useWithMetricsStyle,
+            configureMetricsBuilder: builder => builder.EnableMetrics(meter.Name),
+            configureMeterProviderBuilder: builder => builder.AddInMemoryExporter(exportedItems)))
+        {
+            var counter = meter.CreateCounter<long>("TestCounter");
+            counter.Add(1);
+        }
+
+        AssertSingleMetricWithLongSumOfOne(exportedItems);
+    }
+
+    [Theory]
+    [InlineData(false)]
+    [InlineData(true)]
+    public void EnableMetricsWithAddMeterTest(bool useWithMetricsStyle)
+    {
+        using var meter = new Meter(Utils.GetCurrentMethodName());
+        List<Metric> exportedItems = new();
+
+        using (var host = MetricTestsBase.BuildHost(
+            useWithMetricsStyle,
+            configureMetricsBuilder: builder => builder.EnableMetrics(meter.Name),
+            configureMeterProviderBuilder: builder => builder
+                .AddSdkMeter(meter.Name)
+                .AddInMemoryExporter(exportedItems)))
+        {
+            var counter = meter.CreateCounter<long>("TestCounter");
+            counter.Add(1);
+        }
+
+        AssertSingleMetricWithLongSumOfOne(exportedItems);
+    }
+
+    [Theory]
+    [InlineData(false)]
+    [InlineData(true)]
+    public void ReloadOfMetricsViaIConfigurationTest(bool useWithMetricsStyle)
+    {
+        using var inMemoryEventListener = new InMemoryEventListener(OpenTelemetrySdkEventSource.Log);
+
+        using var meter = new Meter(Utils.GetCurrentMethodName());
+        List<Metric> exportedItems = new();
+
+        var source = new MemoryConfigurationSource();
+        var memory = new MemoryConfigurationProvider(source);
+        var configuration = new ConfigurationRoot(new[] { memory });
+
+        using var host = MetricTestsBase.BuildHost(
+            useWithMetricsStyle,
+            configureAppConfiguration: (context, builder) => builder.AddConfiguration(configuration),
+            configureMeterProviderBuilder: builder => builder
+                .AddInMemoryExporter(exportedItems, reader => reader.TemporalityPreference = MetricReaderTemporalityPreference.Delta));
```

Comment:
This should be tested with Cumulative temporality to ensure that the SDK uses the same `Metric` instance and `MetricPoint[]` to track the measurements.

Reply:
Let me ask you a question first: If we are in cumulative temporality and some instrument is disabled what should happen? Should we continue to report the metric or should we remove it?

I'm not sure what the correct behavior is and when I asked @alanwest he said it is (more or less) undefined.

What actually happens (currently) when an instrument is completed/disposed (regardless of temporality)...

* We report the metric one last time on the next export.
* We remove the metric and no longer report it.
* If it gets reactivated we start reporting it with fresh data.

Why I'm bringing all this up...

Based on your comment it seems you are thinking we should keep the metric around such that when it reactivates we re-use the MetricPoint and continue where we left off. I don't have an issue with that, but we'll have to make some changes to make it so 🚀

Reply:
> Let me ask you a question first: If we are in cumulative temporality and some instrument is disabled what should happen? Should we continue to report the metric or should we remove it?

I think we should continue to report the metric with the original start time of the metric. The consumer of the metric would see a drop when the instrument is disabled but it would still be valuable to them to consume the same metric after they re-enable it. Why should we preclude OTLP/ Prometheus exporter users to make use of the hot-reload feature?

@cijothomas Would love to know your thoughts on this?

> Based on your comment it seems you are thinking we should keep the metric around such that when it reactivates, we re-use the MetricPoint and continue where we left off.

Yes, that's precisely what I'm saying. We have a limited number of metrics that the SDK can report. By default, that is 1000. I don't think we should waste multiple of these `Metric` objects on tracking the same instrument. 

> I don't have an issue with that, but we'll have to make some changes to make it so 🚀

I'm hoping it wouldn't require a lot of changes. Currently, we remove the disposed metric by marking it as `null`. If we don't do that, then we should be able to re-use the same data structure to track the metric any time it gets re-enabled.

Reply:
I suspect that disabling then reenabling should be treated as a [reset or gap](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/data-model.md#resets-and-gaps) - i.e. when reenabled the time series should start over again with a new start time. From what I understand from this discsussion, that is essentially what is happening today.

Reply:
@utpilla @alanwest OK I fixed the bug where the SDK was throwing away and re-creating `Metric` instances. It will now reuse them if something comes back. That actually simplified the logic I had. But I stopped short of resetting the start time and/or clearing the previous measurements. That seemed like a bigger effort that I figured we could tackle independently. I left some TODOs in there with questions I had about where/how to do it.

Reply:
I captured #5064 for the follow up

Reply:
@alanwest FYI @utpilla asked me to revert the fix. So we will continue for now with the behavior of metrics getting removed and recreated.

I did make one change which is to check the flag when deciding whether or not to reuse a metric: https://github.com/open-telemetry/opentelemetry-dotnet/pull/4958/files#diff-8650a8083e99ad454eac9d5bef1a6061cc0c1ab31085a1d25e5c4f57ed16f74fR260-R261

Previous this worked fine:

* Instrument deactivated (metric 1 flagged as inactive)
* Collect runs (metric 1 exported & removed)
* Instrument activated (metric 2 created)

But this...

* Instrument deactivated (metric 1 flagged as inactive)
* Instrument activated (metric 1 reused)
* Collect runs (metric 1 exported & removed)

...would lead to all measurements being lost after the collection. The flag check smooths that out a bit:

* Instrument deactivated (metric 1 flagged as inactive)
* Instrument activated (metric 2 created)
* Collect runs (metric 1 exported & removed, metric 2 exported)

It isn't perfect because you'll get dupes on that first collect but then things will go back to stable on the next collect. That seemed like a better state to be in rather than ghosting the metric.